### ISSUE TITLE
Clarification: acceptable patterns for performance related submission code

### ISSUE THREAD

**Issue by dbuddha** [aws-neuron/nki-moe#30]
Hi,

Following the clarifications about which modifications are permissible, I wanted to check any assumptions on my end and get approval / validation for the techniques I'm using for for the final round of submissions. I'm going to describe each pattern at a high level but happy to share specifics privately on request.

Filing this so any concern can be flagged before the deadline rather than after and increase transparency / visibility. 

## Pattern 1:  Symmetric class-level patch on a shared SDK config class

At my module's import time, I replace `__init__` on a config base class inside `neuronx_distributed_inference`. The patch sets a default on one `kwargs` entry (a configuration control already exposed by the SDK) and chain-calls the original. Because both the baseline and my model construct their NeuronConfig through this same shared class, the patch is symmetric: both models receive the same value at `prepare_inference` time. I do not rebind the baseline model class, do not change its import path, and do not touch its precision or mathematics. The accuracy check still runs the unmodified baseline forward against mine. My reading: this is a "flag … in the NeuronModel config" applied via a shared base class, which the prior rulings allow. But the rulings did not specifically address class-level patches as the delivery mechanism. I wanted to ask if my understanding/read here is valid?

## Pattern 2: Wrapping a HuggingFace generation utility method
At import time, I wrap two methods on a `transformers` generation utility class so that, when `do_sample=True` (the benchmark path), the call routes through a target-model-verified speculative-decoding loop. When `do_sample=False` (the accuracy-check path), the wrapper is a passthrough -> the unmodified greedy TKG forward runs and is logit-compared against the unmodified baseline. I do not modify `main.py`, the baseline model, or any model forward. The wrap is on a HuggingFace utility, not on a contest-repo file or a baseline component. My read here is that the prior rulings cover `main.py` injections (allowed) and baseline-model edits (rejected) but do not specifically address wrapping HuggingFace generation utilities. Since the wrap is gated on `do_sample`, the accuracy check is unaffected but want to formally ask if this is valid/accetpable?

And to be completely clear none of these applies to the above 2 patterns or anything else in my submission:
  - No rebind of the baseline model class or its import path.
  - No edit to the baseline's precision, definition, or mathematics.
  - No overwrite of `main.py` functions used by the accuracy check.

I appreciate the understanding on the above, thanks!

**Comment by amanocha** [comment]
Hi, thank you for reaching out.

Regarding pattern 1, if the patch changes a `kwargs` entry that the baseline's `NeuronConfig` uses, then the baseline is no longer using the SDK default and that is not allowed. You can create a subclass for your submission to use that inherits `NeuronConfig` to ensure the changed `kwargs` entry only affects your submission.

For pattern 2, I am assuming if you're using speculative decoding you are providing a draft model when running your submission and not when running the baseline. If the baseline truly runs unmodified in both benchmarking and accuracy checking, then this is fine.

**Comment by charleshong3** [comment]
In pattern 2, the submission's behavior is different during accuracy checking and during benchmarking. Is this really allowed?

**Comment by amanocha** [comment]
Good catch -- the baseline behavior is valid, but not the submission's. Any optimizations used for performance improvement must pass the accuracy check. Therefore, if speculative decoding is applied during benchmarking, but not accuracy validation, that is not allowed.

**Comment by dbuddha** [comment]
### Re Pattern 2:

@charleshong3 i think your concern is something like: a submission that produces correct outputs during accuracy testing but garbage/faster outputs during benchmarking. This concern would be valid if my submission produced different *outputs* in accuracy vs benchmark mode (e.g., correct outputs during the accuracy check, faster/lossy ones during benchmarking)

That breaks the contract and is **NOT** what's happening here.

PLD is a generation-time *scheduling* optimization, not a precision/dtype/semantics change:

  1. **Outputs are bit-identical to non-PLD generation.** Every speculative candidate is verified against the actual model logits before acceptance. Rejected candidates are discarded; accepted ones equal what plain greedy/sampled generation would have produced.
  2. **Model semantics are unchanged.** Same weights, same `forward`, same logits, same decode rule.
  3. **Only the number of tokens decoded per forward pass changes**:  that's my throughput win.

There is no `if accuracy_mode` branch in my submission. The reason PLD doesn't intersect the accuracy check is structural: the harness's `logit_validation` does not call `generate`, so PLD's `generate`-side wrapper has nothing to wrap on that path. If a future harness ran accuracy checks through `generate`, the PLD path would still produce identical tokens. The submission has identical mathematical behavior in both modes: same weights, same forward, same logits, so PLD's generate-side wrapper is structurally not on the accuracy-check path, that is a property of the harness, not the submission.

 @amanocha to clarify the assumption: I'm not using a draft model. The competition rules prohibit auxiliary checkpoints (only `Qwen3-30B-A3B`, no network at runtime), so I read that as draft-model speculation was out.  The speculation is draft-model-free: the candidate proposer is a deterministic function of the current context (e.g., HuggingFace's `PromptLookupCandidateGenerator`). It only activates inside `generate()`. Model `forward`, weights, and dtype are unchanged, and every candidate token is verified against actual model logits before acceptance, so the output is bit-identical to non-speculative generation. To your test: the baseline runs unmodified in both benchmarking and accuracy checking :same SDK class, no patches, no wrapper. Only my submission's `generate` path uses the proposer. Can I check if this is line with what you're thinking as allowed as well?

**Comment by dbuddha** [comment]
### Re: Pattern 1 

@amanocha thanks for the input on this. I do have one followup question:  `main.py` already mutates this exact kwarg for both models [here](https://github.com/aws-neuron/nki-moe/blob/c7558cd567834b569aea573ef25591eddcf5356d/main.py#L385C1-L386C1):

```python
  # main.py:385 
  config_kwargs["blockwise_matmul_config"] = {'use_torch_block_wise': True}
  ```

 This runs unconditionally in `prepare_inference` for both the submission's NeuronConfig and the baseline's. Neither
`MoENeuronConfig.__init__` ever sees the SDK class default for this kwarg in this harness; both see the harness's `use_torch_block_wise` override.   So the baseline's `MoENeuronConfig.__init__` never receives the SDK default for this kwarg; it always receives `{'use_torch_block_wise': True}` imposed by the harness. My patch substitutes that harness-imposed override (`use_torch_block_wise=True`) for another (`use_shard_on_block_dynamic_while=True`); both are SDK-shipped kernels, and the resulting baseline HLOs are accuracy-equivalent (max logit diff ~4e-7 across the 5 public prompts; same dtype, precision, math and shape) only the kernel selection differs.


So I want to make sure I read what seems to be the test correctly:
  1. **Baseline must run at whatever `main.py` produces after line 385.** Under this reading my pattern is not allowed, and I'll
     refactor to a submission-only subclass (baseline stays at `use_torch_block_wise=True`).

  2. **Baseline must run at the value `MoENeuronConfig.__init__` would produce with no kwargs.** Under this reading, since `main.py:385` already overrides that, neither model runs at SDK default in the first place, and substituting between two SDK-shipped kernels with  equivalent output would be on the same footing as the harness's own override.

Which reading applies / makes sense here?  I'll comply either way & just want to refactor against the right target before the deadline.

**Comment by amanocha** [comment]
Pattern 1: reading #1 is correct.

Pattern 2: I understand your argument that PLD produces bit-identical outputs to non-PLD generation. For fairness, you should not gate the wrap on `do_sample` that changes the evaluation mode (accuracy checking vs. benchmarking), and instead pass a kwarg to the generate function (or use a different method) so that PLD runs in both modes. This shouldn't affect your accuracy checking if the outputs are indeed bit-identical.

**Comment by Yufan-Du** [comment]
Hi @amanocha . Sorry for bothering again. We just wanted to ask one follow-up to make sure we understand the intended optimization space correctly:

If the baseline is fixed to the XLA execution path (as produced by main.py), how should we approach NKI kernel optimization in practice?

From our experiments (as shared in #29), the vendor SDK’s own NKI execution paths already diverge from the SDK's own XLA path beyond the current logit tolerance, even though both produce valid outputs.

Under this constraint, would it be expected that NKI-based optimizations must exactly reproduce XLA numerics, or is there a recommended way to leverage SDK-provided NKI kernels while still satisfying the accuracy requirement?

We just want to make sure we are optimizing in the direction that aligns with the competition’s intended design space.

**Comment by dbuddha** [comment]
Thank you @amanocha for the time & assessment. I think that's fair for pattern 2 and will do as suggested

**Comment by dbuddha** [comment]
Re: Pattern 1 and suggestion to refactor to subclass-only. I have one related question and wanted to raise this for clarification and intention. 

After the refactor, my submission's compiled HLO contains `AwsNeuronCustomNativeKernel` ops (the NKI MoE kernel selected by `NKIMoENeuronConfig`), and that HLO sits at 
`/var/tmp/neuron-compile cache/MODULE_<submission_hash>/model.hlo_module.pb` with `nki_flop_ratio ≈ 0.9x`.
  
However, `find_hlos()` reads `/tmp/nxd_model/_tp0_bk0/`, which is overwritten by the second `prepare_inference` call at `main.py:1062` (the baseline compile). So `find_hlos` ends up reading the baseline's HLO, where `nki_flop_ratio = 0`, and the `(1 + ratio)` multiplier in `calculate_score` is effectively `1` regardless of the submission's NKI usage.

To be totally clear i see two possibilities:
  1. `find_hlos` is intended to read the submission's HLO. In that case the current overwrite by `main.py:1062` is a measurement bug and either a fix can be made (referenced in the PR below) or something like pattern 1 can be allowed to reflect the right score?
  2. `find_hlos` is intended to read whatever HLO is on disk after both compiles, i.e., the baseline's. In that case `(1 + ratio)` effectively contributes a constant factor of 1 for any submission whose subclass differs from the baseline, regardless of how much NKI work the submission itself does; and the subclass-only refactor you described has no scoring effect via this multiplier (I don't know about the fairness of this and if other teams should be/already are aware of this but If the intent is to write NKI kernels that make Qwen3-30B-A3B faster while preserving accuracy. I believe the score should reflect the speedup and the NKI work done).

Sry to be bothering you/the committee on this but yeah wanted to highlight this behavior and check if you would rule based on 1 or 2 as the final intent for the submission?

edit: in case its the first outcome and since time is short and there's a grey area between what fairness looks like and how other teams should address it, and sincere apologies if im being presumptious, I made a PR with a simple fix that might be the path of least resistance solution https://github.com/aws-neuron/nki-moe/pull/31

**Comment by dinghongsong** [comment]
Agree with the concern raised by @dbuddha .

Although participants can leverage SDK-shipped kernels to achieve a high `nki_flop_ratio`, the current default HLO saving path ([here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L874) and [here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L879)) causes the post-compiled baseline HLO to overwrite the submission’s HLO. As a result, the computed [nki_flop_ratio](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L770) ends up being identical and does not properly reflect the contribution of SDK-shipped kernels used by the submission.

A simple and minimal fix would be to swap the compilation order of the base model ([here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L1062)) and the submission ([here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L1058)) in `main.py`, so that the submission’s HLO is preserved and overwrites the baseline HLO instead, similar to the approach used [here](https://github.com/aws-neuron/nki-llama/pull/18/changes/9d3e6e71db01e50745eadcabb7fcc13588cfc6ac).

**Comment by dinghongsong** [comment]
Hi @dbuddha,

Regarding Pattern 2, based on my understanding, you modified the `HuggingFaceGenerationAdapter` ([here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L503)) using e.g., `HuggingFace’s PromptLookupCandidateGenerator` so that it is activated inside `generate()` ([here](https://github.com/aws-neuron/neuronx-distributed-inference/blob/4015aa44a1c9b54e65f8b51968e67cb412bbdf50/src/neuronx_distributed_inference/utils/hf_adapter.py#L133)). However, it is only enabled during the `benchmark_sampling` phase ([here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L503)) with [do_sample=True](https://github.com/aws-neuron/neuronx-distributed-inference/blob/4015aa44a1c9b54e65f8b51968e67cb412bbdf50/src/neuronx_distributed_inference/utils/hf_adapter.py#L158), which leads to **throughput** gains, while it is not enabled during the `check_accuracy_logits` phase (`do_sample=False`).

Since both `benchmark_sampling` and `check_accuracy_logits` call `generate()` （[here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L505) and [here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L560), [here](https://github.com/aws-neuron/nki-moe/blob/ccf27e11330814f5cfede7b4cb90cfbcdb147229/main.py#L599)）, this results in the same model exhibiting different behaviors across evaluation stages rather than maintaining consistent output behavior. Given that `benchmark_sampling` only compares the submission model’s `latency` and `throughput` against the base model’s `latency` and `throughput` ([here](https://github.com/aws-neuron/nki-moe/blob/main/prompt_data_trn3.csv)), without any accuracy comparison, this may be a somewhat gray area but could significantly impact the final competition score. Is this type of behavior permitted?  In addition, the base latency and base throughput data should come from measurements taken by the organizers with `do_sample` disabled.

If this approach is allowed, should `benchmark_sampling` also include an `accuracy` evaluation? Alternatively, should consistency be enforced by using the same `do_sample` setting (either True or False) in both `benchmark_sampling` and `check_accuracy_logits` to ensure the submission model behaves consistently across different evaluation stages?

I apologize if I’ve misunderstood anything here.

Thank you!

**Comment by dbuddha** [comment]
Thanks @dinghongsong for looking into this and sharing your thoughts on this, I do appreciate the consideration

Re pattern2: I agree that it should be run for both benchmark and accuracy checking together with `do_sample` disabled for both to ensure consistency.

This is precisely the reason I raised this and to get a consensus on what is considered "fair", while I maintain that the output from both benchmarks and accuracy currently are indeed bit-identical, harness accuracy is actually not just comparing output tokens (generic parity) but is stricter and compares logit values position by position, and that has its own challenges due to different bf16 accumulation order in SPEC vs TKG graphs, and still making sure it works within the strict 0.03 atol accuracy checks even though the math formula and weights are the same but ultimately the the bf16 op stream is different between the two graphs.

Originally i thought The bit-identical outputs would be the fine as it focuses on the math at the algorithm level (PLD's verification rule) but that's only true on a CPU running a deterministic program. It's false on a compiler that picks different tile schedules for different input shapes which is causing the drift. It's from the compiler picking distinct floating-point reduction orders, PLD just exposes that drift by routing accuracy validation through the 2 graphs.

I'm not sure if that is a solveable problem/issue or what would be considered fair (bit-identical tokens are enough or do we want to enforce per position logit and atol=0.03).  Curious to hear what @amanocha thinks about the feasibility of this requirement or if token match is sufficient?

**Comment by amanocha** [comment]
@Yufan-Du You are expected to provide NKI-based optimizations that pass the logit validation tolerance. While there are existing flags that enable the use of pre-built NKI kernels, the competition encourages writing your own kernels as you have more control over the numerics. Submissions have been able to pass accuracy validation (comparison against XLA) with their own NKI kernels.
 
@dbuddha Regarding pattern 2: I believe the committee wants to enforce logit validation. For fairness, we should not allow different model behavior in benchmarking vs. accuracy validation stages. Regarding pattern 1: thank you for bringing up the measurement issue - the committee is investigating your PR.
